### PR TITLE
Add Plex Connect Locally

### PR DESCRIPTION
**Pull request by lordbah**
Add Plex Connect Locally
Set Favorite for tracks rated 5 stars in Plex

### CODE REVIEW

**Comment by lordbah** [comment]
When adding the Plex music provider you can avoid using Myplex.TV. Fill in the IP address and port and click Connect Locally. There is no visible reaction. Hit Save. The provider syncs in the background, connecting directly to the Plex server without requiring a token from Myplex.TV. Also, tracks rated 5 stars in Plex will have the Favorite flag set.

File: music_assistant/server/models/music_provider.py [hunk 1]
```diff
@@ -410,7 +410,8 @@ async def sync_library(self, media_types: tuple[MediaType, ...]) -> None:
                         # create full db item
                         # note that we skip the metadata lookup purely to speed up the sync
                         # the additional metadata is then lazy retrieved afterwards
-                        prov_item.favorite = True
+                        if (self.is_streaming_provider):
```
Comment (marcelveldt) [inline music_assistant/server/models/music_provider.py:413]:
```suggestion
                        if self.is_streaming_provider:
```

**Comment by marcelveldt** [comment]
I made a few changes to the PR to make it more explicit that you can choose between Local auth and MyPlex auth.
I had no way to test the local auth option, can give that a double check before we merge here ?

**Comment by marcelveldt** [comment]
Can you deliver a small piece of text for the documentation about using the local auth ?

**Review by marcelveldt — DISMISSED** [review]

**Comment by lordbah** [comment]
> Can you deliver a small piece of text for the documentation about using the local auth ?

Is the comment in the PR sufficient?

**Comment by marcelveldt** [comment]
> > Can you deliver a small piece of text for the documentation about using the local auth ?
> 
> Is the comment in the PR sufficient?

Uhh well, I tried it myself and it didnt work out of the box so I assume there is also a part needed in Plex itself ?

File: music_assistant/server/providers/plex/__init__.py [hunk 2]
```diff
@@ -413,10 +492,14 @@ async def _parse_album(self, plex_album: PlexAlbum) -> Album:
...
                     provider_instance=self.instance_id,
-                    url=plex_album.getWebURL(),
+                    url=plex_album.getWebURL(self._baseurl),
                 )
             },
         )
+        # Only add 5-star rated albums to Favorites. rating will be 10.0 for those.
+        # TODO: Let user set threshold?
+        # album.favorite = True if plex_album._data.attrib['rating'] == "10.0" else False
+
```
Comment (marcelveldt) [inline music_assistant/server/providers/plex/__init__.py:502]: This is outdated ? Or another reason why its commented out ?

**Comment by lordbah** [comment]
> > > Can you deliver a small piece of text for the documentation about using the local auth ?
> > 
> > 
> > Is the comment in the PR sufficient?
> 
> Uhh well, I tried it myself and it didnt work out of the box so I assume there is also a part needed in Plex itself ?

In Plex under Settings / Remote Access, "List of IP addresses and networks that are allowed without auth" is a comma-separated list. Mine is set to "192.168.218.0/255.255.254.0,192.168.220.165".

File: music_assistant/server/providers/plex/__init__.py (hunk 2 above)
Comment (lordbah) [reply music_assistant/server/providers/plex/__init__.py:503]: Oh yeah. I hadn't actually rated any of my albums in Plex, only tracks. Suppose I can mark a few and test that.

**Comment by marcelveldt** [comment]
> > > > Can you deliver a small piece of text for the documentation about using the local auth ?
> > > 
> > > 
> > > Is the comment in the PR sufficient?
> > 
> > 
> > Uhh well, I tried it myself and it didnt work out of the box so I assume there is also a part needed in Plex itself ?
> 
> In Plex under Settings / Remote Access, "List of IP addresses and networks that are allowed without auth" is a comma-separated list. Mine is set to "192.168.218.0/255.255.254.0,192.168.220.165". I guess I assumed that anyone who would want to connect locally would already have that.

Let's add this info to the docs then as it was new info for me. Thanks!

File: music_assistant/server/providers/plex/__init__.py (hunk 2 above)
Comment (lordbah) [reply music_assistant/server/providers/plex/__init__.py]: Done.

File: music_assistant/server/providers/plex/__init__.py [hunk 3]
```diff
@@ -526,10 +614,17 @@ async def _parse_track(self, plex_track: PlexTrack) -> Track:
...
                 )
             },
         )
+        # Only add 5-star rated tracks to Favorites. userRating will be 10.0 for those.
+        # TODO: Let user set threshold?
+        try:
+            track.favorite = plex_track._data.attrib["userRating"] == "10.0"
+        except KeyError:
+            # Log but suppress exception, allow sync to continue
+            self.logger.error(f"ERROR {plex_track.title} has no userRating")
```
Comment (marcelveldt) [inline music_assistant/server/providers/plex/__init__.py:627]:
```suggestion
            self.logger.error("ERROR: %s has no userRating", plex_track.title)
```

Comment (marcelveldt) [reply music_assistant/server/providers/plex/__init__.py:627]: Don't use f-string formatting in log strings

File: music_assistant/server/providers/plex/__init__.py [hunk 4]
```diff
@@ -413,10 +493,18 @@ async def _parse_album(self, plex_album: PlexAlbum) -> Album:
...
                 )
             },
         )
+        # Only add 5-star rated albums to Favorites. rating will be 10.0 for those.
+        # TODO: Let user set threshold?
+        try:
+            album.favorite = plex_album._data.attrib["userRating"] == "10.0"
+        except KeyError:
+            # Log but suppress exception, allow sync to continue
+            self.logger.error(f"ERROR {plex_album.title} has no rating")
```
Comment (marcelveldt) [inline music_assistant/server/providers/plex/__init__.py:506]:
```suggestion
            self.logger.error("ERROR: %s has no rating", plex_album.title)
```

**Review by marcelveldt — APPROVED** [review]
Thanks @lordbah !

**Pull request merged**